sdk/core/azure-core/azure/core/exceptions.py

```diff
@@ -158,11 +157,11 @@ class ODataV4Format(object):
     def __init__(self, json_object):
         if "error" in json_object:
             json_object = json_object["error"]
-        cls = self.__class__  # type: Type[ODataV4Format]
+        cls: Type[ODataV4Format] = self.__class__
 
         # Required fields, but assume they could be missing still to be robust
-        self.code = json_object.get(cls.CODE_LABEL)  # type: Optional[str]
-        self.message = json_object.get(cls.MESSAGE_LABEL)  # type: Optional[str]
+        self.code: Optional[str] = json_object.get(cls.CODE_LABEL)
```

Comment:
For public instance attributes, I would suggest adding them as class-level type annotations (per PEP526). But (as always), we have to check the generated docs to make sure we don't break sphinx and friends... @kristapratico may also have opinions on this...

Reply:
It's not required: 

![image](https://user-images.githubusercontent.com/1050156/210607046-10ae56c5-0606-4d85-bc27-d474e60e25fc.png)

But I will check the impact it would had on doc, I'm curious

Reply:
One thing to consider is that if we don't type the parameters in the init signature, we don't get type checking when the obj is created: https://mypy-play.net/?mypy=latest&python=3.11&gist=19416a26bbfbbb641f46d0383d238844 (pyright does the same thing here). If we expect a user to create the obj, I think we should definitely type the init signature and/or maybe for ourselves to catch bugs.

I didn't have a strong opinion on public instance attributes being typed via PEP526 or inside the init body, but I think the documentation experience is what pushes me towards the former. We get a lot more with class-level type hints whereas the init body hints don't add anything to the docs.

Note: using the newer sphinx v5.3.0 in these screenshots since we plan to bump to this version soon. With class-level type hints we get nice docs and side-bar documentation of the public attrs:

![image](https://user-images.githubusercontent.com/31998003/210635672-6b2f1d70-7f2b-4bcd-a3c4-734518eb43b6.png)

For comparison, init body hints don't get us the same info:

![image](https://user-images.githubusercontent.com/31998003/210636885-a6edb853-bfd4-4c2a-9bf1-f8500c22c223.png)


If we go with the class body hints (PEP526), the docs get a bit redundant with the docstring explanation of ivars so maybe we could remove ivar docstrings like below. Removing the docstring info changes the hover-over experience in IDEs so not totally sure that we'd want to do that, but it does feel a lot cleaner.

![image](https://user-images.githubusercontent.com/31998003/210635105-ddb5efb3-3ea1-403b-8904-be7011c6a02f.png)

Reply:
However, for MS learn docs, the class-level hints don't seem to render as nice (shown as Attributes): https://learn.microsoft.com/en-us/python/api/azure-ai-anomalydetector/azure.ai.anomalydetector.models.alignpolicy?view=azure-python-preview

That was created with the older sphinx version so not sure if it will look better with v5.3.0.

Reply:
Thanks for the extensive testing! Those are attributes, this is not wrong actually on learn doc. I was afraid of worse to be honest (like no show at all).

And yes I would remove ivar once we go this way, and keep the param only.

Reply:
Oki, I fully typed CloudEvent to follow the new convention @johanste , curious to get your thoughts